adam-core/src/main/scala/org/bdgenomics/adam/util/IndexedFastaFile.scala

```diff
@@ -0,0 +1,82 @@
+/**
+ * Licensed to Big Data Genomics (BDG) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The BDG licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bdgenomics.adam.util
+
+import java.net.URI
+import java.nio.file.Paths
+
+import htsjdk.samtools.reference.{ FastaSequenceIndex, IndexedFastaSequenceFile }
+import org.apache.hadoop.fs.Path
+import org.apache.spark.SparkContext
+import org.bdgenomics.adam.models.{ SequenceDictionary, ReferenceRegion }
+import org.bdgenomics.utils.misc.Logging
+import picard.sam.CreateSequenceDictionary
+
+/**
+ * Loads and extracts sequences directly from indexed fasta or fa files. filePath requires fai index in the
+ * same directory with same naming convention.
+ *
+ * @param filePath path to fasta or fa index
+ */
+case class IndexedFastaFile(sc: SparkContext, filePath: String) extends ReferenceFile with Logging {
+
+  // get absolute path and scheme to create URI
+  val path = new Path(filePath).toString
+  val scheme: String = new Path(path).getFileSystem(sc.hadoopConfiguration).getScheme
+  val pathWithScheme = s"${scheme}://${path}"
+
+  val uri: URI = new URI(pathWithScheme)
+
+  val file = Paths.get(uri).toFile
+
+  // Generate IndexedFastaSequenceFile from path and fai index
+  val ref: IndexedFastaSequenceFile = {
+    val uriIdx: URI = new URI(pathWithScheme + ".fai")
+    val pathIdx = Paths.get(uriIdx)
+
+    val idx = new FastaSequenceIndex(pathIdx.toFile)
+    new IndexedFastaSequenceFile(file, idx)
+  }
+
+  // Get sequence dictionary. If sequence dictionary is not defined,
+  // generate sequence dictionary from dile
+  val sequences =
+    try {
+      SequenceDictionary(ref.getSequenceDictionary)
+    } catch {
+      case e: Exception => {
```

Comment:
I'd rather not generate the Sequence Dictionary file, my preference would be to fall back on a validation stringency check here. I'm thinking it'd be best to:

```
} catch {
  case e: Exception => {
    if (stringency == ValidationStringency.STRICT) {
      throw e
    } else {
      if (stringency == ValidationStringency.LENIENT) {
        log.warn("Caught exception %s when loading FASTA sequence dictionary. Using empty dictionary instead.".format(e))
      }
      SequenceDictionary.empty
    }
  }
}
```